libraries/Microsoft.Bot.Builder.Ai/LocaleConverterMiddleware.cs

```diff
@@ -48,22 +49,44 @@ public LocaleConverterMiddleware(Func<ITurnContext, string> getUserLocale, Func<
             {
                 if (!String.IsNullOrWhiteSpace(message.Text))
                 {
-                    string fromLocale = _getUserLocale(context);
-                    ConvertLocaleMessage(message, fromLocale);
-                    await _setUserLocale(context);
+                    bool localeChanged = await _setUserLocale(context);
+                    if (!localeChanged)
+                    {
+                        string fromLocale = _getUserLocale(context);
+                        ConvertLocaleMessage(context, fromLocale);
+                    }
+                    
                 }
             }
-            await next().ConfigureAwait(false);
+            if(!_isLastMiddleware)
+                await next().ConfigureAwait(false);
         }
 
-        private void ConvertLocaleMessage(IMessageActivity message,string fromLocale)
+        private void ConvertLocaleMessage(ITurnContext context,string fromLocale)
         {
-            
-            if (localeConverter.IsLocaleAvailable(fromLocale) && fromLocale != toLocale)
+            IMessageActivity message = context.Activity.AsMessageActivity();
```

Comment:
This could return null. If it does, the code below may not work correctly. Consider a null-check, or an Assertion around the activity type.

Reply:
Updated